src/api/exchange/ExchangeApi.ts

```diff
@@ -101,21 +114,156 @@ export interface Order {
   remainingAmount: BN
 }
 
+/**
+ * BaseTradeEvent uses only info available on the emitted Trade event
+ */
+export interface BaseTradeEvent {
+  // order related
+  orderId: string
+  sellTokenId: number
+  buyTokenId: number
+  sellAmount: BN
+  buyAmount: BN
+  // block related
+  txHash: string
+  txIndex: number
+  blockNumber: number
+  id: string // txHash + | + txIndex
+}
+
+/**
+ * TradeEvent enriches BaseTradeEvent with block, order and token data
+ */
+export interface TradeEvent extends BaseTradeEvent {
+  batchId: number
+  hashKey: string // orderId + batchId, to find reverts
+  indexOnBatch: number // tracks trade position on batch, in case of reverts
+  time: Date
+  buyToken: TokenDetails
+  sellToken: TokenDetails
+  limitPrice: BigNumber
+  fillPrice: BigNumber
+  reverted: boolean
+}
+
 export interface GetOrdersPaginatedResult {
   orders: AuctionElement[]
   nextIndex?: number
 }
 
+// TODO: move to const/config
+const CONTRACT_DEPLOYMENT_BLOCK = {
+  1: 9340147,
+  4: 5844678,
+}
+
+interface Subscriptions {
+  // TODO: didn't manage to infer this type from the contract. How can I do that?
+  trade?: ContractEventEmitter<{
+    owner: string
+    orderId: string
+    sellToken: string
+    buyToken: string
+    executedSellAmount: string
+    executedBuyAmount: string
+  }>
+}
 /**
  * Basic implementation of Stable Coin Converter API
  */
 export class ExchangeApiImpl extends DepositApiImpl implements ExchangeApi {
+  private subscriptions: Subscriptions = {}
+
   public constructor(injectedDependencies: Params) {
     super(injectedDependencies)
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
     ;(window as any).exchange = this._contractPrototype
   }
 
+  public async getPastTrades({ userAddress, networkId }: PastEventsParams): Promise<BaseTradeEvent[]> {
+    const contract = await this._getContract(networkId)
+
+    // to get all past events
+    const tradeEvents = await contract.getPastEvents('Trade', {
```

Comment:
Not typed by **typechain**, but we can and probably should type it ourselves in dex-js
I'll look into it